File: internal/engine/manifest.go
package engine

import (
	"bufio"
	"encoding/json"
	"os"
	"sync"
)

type ManifestRecord struct {
	Action string `json:"action"`
	Path   string `json:"path"`
	MinKey string `json:"min_key"`
	MaxKey string `json:"max_key"`
}

type Manifest struct {
	file *os.File
	mu   sync.Mutex
	path string
}

func NewManifest(path string) (*Manifest, error) {
	f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return nil, err
	}
	return &Manifest{
		file: f,
		path: path,
	}, nil
}

func (m *Manifest) Append(record ManifestRecord) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	data, err := json.Marshal(record)
	if err != nil {
		return err
	}

	data = append(data, '\n')
	if _, err := m.file.Write(data); err != nil {
		return err
	}

	return m.file.Sync()
}

func (m *Manifest) Compact(activeRecords []ManifestRecord) error {
	m.mu.Lock()
	defer m.mu.Unlock()

	tempPath := m.path + ".tmp"
	tempFile, err := os.OpenFile(tempPath, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
	if err != nil {
		return err
	}

	for _, rec := range activeRecords {
		rec.Action = "ADD"
		data, _ := json.Marshal(rec)
		data = append(data, '\n')
		tempFile.Write(data)
	}

	tempFile.Sync()
	tempFile.Close()

	m.file.Close()

	if err := os.Rename(tempPath, m.path); err != nil {
		return err
	}

	f, err := os.OpenFile(m.path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		return err
	}
	m.file = f

	return nil
}

func (m *Manifest) Close() error {
	m.mu.Lock()
	defer m.mu.Unlock()
	return m.file.Close()
}

func RecoverManifest(path string) ([]ManifestRecord, error) {
	f, err := os.Open(path)
	if err != nil {
		if os.IsNotExist(err) {
			return []ManifestRecord{}, nil
		}
		return nil, err
	}
	defer f.Close()

	var records []ManifestRecord

	isDead := make(map[int]bool)
	pathToIdx := make(map[string]int)

	scanner := bufio.NewScanner(f)
	idx := 0

	for scanner.Scan() {
		var rec ManifestRecord
		if err := json.Unmarshal(scanner.Bytes(), &rec); err != nil {
			break
		}

		if rec.Action == "ADD" {
			records = append(records, rec)
			pathToIdx[rec.Path] = idx
			idx++
		} else if rec.Action == "REMOVE" {
			if targetIdx, exists := pathToIdx[rec.Path]; exists {
				isDead[targetIdx] = true
			}
		}
	}

	var activeTables []ManifestRecord
	for i, rec := range records {
		if !isDead[i] {
			activeTables = append(activeTables, rec)
		}
	}

	return activeTables, nil
}

File: internal/engine2/Manifest.cpp
#include "Manifest.hpp"
#include <filesystem>
#include <unordered_map>
#include <sstream>

Manifest::Manifest(const std::string& manifestPath) : path(manifestPath) {
    file.open(path, std::ios::out | std::ios::app | std::ios::binary);
}

Manifest::~Manifest() {
    Close();
}

std::string Manifest::serialize(const ManifestRecord& record) {
    // Generates raw JSON matching Go's exact serialization format
    return "{\"action\":\"" + record.Action + 
           "\",\"path\":\"" + record.Path + 
           "\",\"min_key\":\"" + record.MinKey + 
           "\",\"max_key\":\"" + record.MaxKey + "\"}";
}

bool Manifest::Append(const ManifestRecord& record) {
    std::lock_guard<std::mutex> lock(mu);
    if (!file.is_open()) return false;

    std::string data = serialize(record) + "\n";
    file.write(data.data(), data.size());
    file.flush(); // Equates to Go's file.Sync() for stream flushing
    
    return file.good();
}

bool Manifest::Compact(const std::vector<ManifestRecord>& activeRecords) {
    std::lock_guard<std::mutex> lock(mu);

    std::string tempPath = path + ".tmp";
    std::ofstream tempFile(tempPath, std::ios::out | std::ios::trunc | std::ios::binary);
    if (!tempFile.is_open()) return false;

    for (auto rec : activeRecords) {
        rec.Action = "ADD"; // Force compaction history to clean state additions
        std::string data = serialize(rec) + "\n";
        tempFile.write(data.data(), data.size());
    }

    tempFile.flush();
    tempFile.close();

    if (file.is_open()) {
        file.close();
    }

    std::error_code ec;
    std::filesystem::rename(tempPath, path, ec);
    if (ec) return false;

    file.open(path, std::ios::out | std::ios::app | std::ios::binary);
    return file.is_open();
}

void Manifest::Close() {
    std::lock_guard<std::mutex> lock(mu);
    if (file.is_open()) {
        file.close();
    }
}

// Robust custom JSON stream parser to avoid external dependencies
static bool DeserializeRecord(const std::string& line, ManifestRecord& rec) {
    auto extractField = [](const std::string& json, const std::string& key) -> std::string {
        std::string pattern = "\"" + key + "\":\"";
        size_t startPos = json.find(pattern);
        if (startPos == std::string::npos) return "";
        startPos += pattern.length();
        
        size_t endPos = json.find("\"", startPos);
        if (endPos == std::string::npos) return "";
        
        return json.substr(startPos, endPos - startPos);
    };

    rec.Action = extractField(line, "action");
    rec.Path = extractField(line, "path");
    rec.MinKey = extractField(line, "min_key");
    rec.MaxKey = extractField(line, "max_key");

    return !rec.Action.empty() && !rec.Path.empty();
}

std::vector<ManifestRecord> RecoverManifest(const std::string& path) {
    if (!std::filesystem::exists(path)) {
        return {};
    }

    std::ifstream f(path, std::ios::in | std::ios::binary);
    if (!f.is_open()) {
        return {};
    }

    std::vector<ManifestRecord> records;
    std::unordered_map<std::string, size_t> pathToIdx;
    std::unordered_map<size_t, bool> isDead;

    std::string line;
    size_t idx = 0;

    while (std::getline(f, line)) {
        if (line.empty()) continue;
        
        ManifestRecord rec;
        if (!DeserializeRecord(line, rec)) {
            break; 
        }

        if (rec.Action == "ADD") {
            records.push_back(rec);
            pathToIdx[rec.Path] = idx;
            idx++;
        } else if (rec.Action == "REMOVE") {
            if (pathToIdx.find(rec.Path) != pathToIdx.end()) {
                size_t targetIdx = pathToIdx[rec.Path];
                isDead[targetIdx] = true;
            }
        }
    }
    f.close();

    std::vector<ManifestRecord> activeTables;
    for (size_t i = 0; i < records.size(); ++i) {
        if (isDead.find(i) == isDead.end()) {
            activeTables.push_back(records[i]);
        }
    }

    return activeTables;
}